Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -655,13 +637,46 @@ void CGPathAddRoundedRect(
     UNIMPLEMENTED();
 }
 
+int _CGPathPointCountForElementType(CGPathElementType type) {
+    int pointCount = 0;
+
+    switch (type) {
+        case kCGPathElementMoveToPoint:
+        case kCGPathElementAddLineToPoint:
+            pointCount = 1;
+            break;
+        case kCGPathElementAddQuadCurveToPoint:
+            pointCount = 2;
+            break;
+        case kCGPathElementAddCurveToPoint:
+            pointCount = 3;
+            break;
+        case kCGPathElementCloseSubpath:
+            pointCount = 0;
+            break;
+    }
+    return pointCount;
+}
+
 /**
  @Status Interoperable
 */
 void CGPathApply(CGPathRef path, void* info, CGPathApplierFunction function) {
+    if (path == NULL) {
+        return;
+    }
+
     for (unsigned i = 0; i < path->_count; i++) {
-        CGPathElement c = path->_components[i];
-        function(info, &c);
+        CGPathElement element = path->_components[i];
+        CGPathElement returnElement;
+        returnElement.type = element.type;
+        int pointCount = _CGPathPointCountForElementType(element.type);
+        returnElement.points = (CGPoint*)IwCalloc(pointCount, sizeof(CGPoint));
+        memcpy(returnElement.points, element.points, pointCount * sizeof(CGPoint));
```

Comment:
I agree that anyone trying to manipulate paths is using this API incorrectly. However, if I don't discard changes, then this does become a spot where users can intentionally or unintentionally manipulate paths. If that is the case, the applier function would be behaving differently from the documentation as a function that "can examine (but not modify) the element" because it would be able to modify the element.